telegram_notifier.py

```diff
@@ -0,0 +1,125 @@
+    def __init__(self):
+        self.bot_token = os.getenv("TELEGRAM_BOT_TOKEN")
+        self.chat_id = os.getenv("TELEGRAM_CHAT_ID")
+
+        if not self.bot_token or not self.chat_id:
+            raise ValueError("Missing TELEGRAM_BOT_TOKEN or TELEGRAM_CHAT_ID in .env file")
+
+        self.api_url = f"https://api.telegram.org/bot{self.bot_token}/sendMessage"
+    
+        self.emoji_scale = [
+                                "😞",  # 0–2
+                                "😐",  # 2–4
+                                "😌",  # 4–6
+                                "🙂",  # 6–8
+                                "😃",  # 8–10
+                            ]
+    
+    async def score_to_emoji(self, score: float) -> str:
+        if score < 2:
+            return self.emoji_scale[0]
+        elif score < 4:
+            return self.emoji_scale[1]
+        elif score < 6:
+            return self.emoji_scale[2]
+        elif score < 8:
+            return self.emoji_scale[3]
+        else:
+            return self.emoji_scale[4]
+
+    async def format_message(self, data: dict) -> str:
+        pr = TelegramOutputSending(**data)
+
+        message = (
+            f"<b>📦 Pull Request Score Report</b>\n"
+            f"👤 <b>Committer:</b> <code>{pr.committer}</code>\n"
+            f"📁 <b>Repo:</b> <code>{pr.repo}</code>\n"
+            f"🔢 <b>PR Number:</b> <code>#{pr.pr_number}</code>\n\n"
+            f"🔗 <b>Commit:</b> <code>{pr.commit}</code> 🛠️\n\n"
+            f"<b>📊 Detail of each criterion:</b>\n"
+        )
+
+        for c in pr.criteria:
+            emoji = await self.score_to_emoji(c.score)
+            message += f"• <b>{c.name}:</b> <code>{c.score:.1f}</code> {emoji}\n"
+
+        total_emoji = await self.score_to_emoji(pr.total_score)
+
+        message += (
+            f"\n<b>🌟 Total Score:</b> <code>{pr.total_score:.1f}/10</code> {total_emoji}\n\n"
+            f"<b>🧾 Overall Comment:</b>\n<i>{pr.final_comment}</i>"  # No escaping needed for _
+        )
+
+        return message
+
+    # async def format_message(self, data: dict) -> str:
+    #     pr = TelegramOutputSending(**data)
+
+    #     message = (
+    #         f"*📦 Pull Request Score Report*\n"
+    #         f"👤 *Committer:* `{pr.committer}`\n"
+    #         f"📁 *Repo:* `{pr.repo}`\n"
+    #         f"🔢 *PR Number:* `#{pr.pr_number}`\n\n"
+    #         f"🔗 *Commit:* `{pr.commit}` 🛠️\n\n"
+    #         f"*📊 Chi tiết điểm theo từng tiêu chí:*\n"
+    #     )
+
+    #     for c in pr.criteria:
+    #         emoji = await self.score_to_emoji(c.score)
+    #         message += f"• *{c.name}*: `{c.score:.1f}` {emoji}\n"
+
+    #     total_emoji = await self.score_to_emoji(pr.total_score)
+
+    #     message += (
+    #         f"\n*🌟 Tổng điểm:* `{pr.total_score:.1f}/10` {total_emoji}\n\n"
+    #         f"*🧾 Nhận xét tổng quát:*\n_{pr.final_comment}_"
+    #     )
+
+
+    #     return message
+
+    async def send_message(self, message: str) -> None:
```

Comment:
High severity: send_message is declared async but makes blocking HTTP POST requests with requests.post, which blocks the event loop and harms async performance. Replace requests with an async HTTP client such as aiohttp or httpx.AsyncClient, add timeout handling and robust exception management to avoid blocking and ensure reliable message delivery. Example refactor:
```suggestion
import aiohttp

async def send_message(self, message: str):
    async with aiohttp.ClientSession() as session:
        try:
            async with session.post(self._url, json={'text': message, 'parse_mode': 'HTML'}, timeout=10) as resp:
                resp.raise_for_status()
        except Exception as e:
            # handle/log exception
            raise
``` (Severity: high, score: 9/10)